lib/Translation/ExportVerilog/ExportVerilog.cpp

```diff
@@ -159,6 +161,106 @@ static bool isZeroBitType(Type type) {
+  TypeSwitch<Type, void>(type)
+      .Case<InOutType>([&](InOutType inout) {
+        printArraySubscriptsPre(inout.getElementType(), loc, os);
+      })
+      .Case<ArrayType>([&](ArrayType array) {
+        os << '[' << (array.getSize() - 1) << ":0]";
+        printArraySubscriptsPre(array.getElementType(), loc, os);
+      })
+      .Case<UnpackedArrayType>([&](UnpackedArrayType array) {
+        printArraySubscriptsPre(array.getElementType(), loc, os);
+      })
+      .Case<IntegerType>([&](IntegerType integer) {
+        if (integer.getWidth() == 0)
+          os << "/*Zero Width*/";
+        else if (integer.getWidth() != 1)
+          os << '[' << (integer.getWidth() - 1) << ":0]";
+      })
+      .Case<StructType>([&](StructType) {})
+      .Default([&](Type t) {
+        mlir::emitError(loc, "value has an unsupported verilog type ") << t;
+      });
+}
+
+// Print out the unpacked array subscripts after a variable/port/field name.
+static bool hasArraySubscriptsPre(Type type) {
+  return TypeSwitch<Type, bool>(type)
+      .Case<InOutType>([&](InOutType inout) {
+        return hasArraySubscriptsPre(inout.getElementType());
+      })
+      .Case<ArrayType>([&](ArrayType array) { return true; })
+      .Case<UnpackedArrayType>([&](UnpackedArrayType array) {
+        return hasArraySubscriptsPre(array.getElementType());
+      })
+      .Case<IntegerType>(
+          [&](IntegerType integer) { return integer.getWidth() != 1; })
+      .Default([&](Type) { return false; });
+}
+
+// Print out the array subscripts after a wire/port declaration.
+static void printArraySubscriptsPost(Type type, raw_ostream &os) {
+  if (auto inout = type.dyn_cast<InOutType>())
+    printArraySubscriptsPost(inout.getElementType(), os);
+  else if (auto array = type.dyn_cast<UnpackedArrayType>()) {
+    printArraySubscriptsPost(array.getElementType(), os);
+    os << '[' << (array.getSize() - 1) << ":0]";
+  }
+}
+
+// Output the basic type that is to the left of any dimensions
+// returns whether a base type was emitted
+static bool baseType(Type type, StringRef baseIntType, Location loc,
+                     raw_ostream &os) {
+  return TypeSwitch<Type, bool>(type)
+      .Case<IntegerType>([&](IntegerType integerType) {
+        if (baseIntType.empty())
+          return false;
+        os << baseIntType;
+        return true;
+      })
+      .Case<InOutType>([&](InOutType inoutType) {
+        return baseType(inoutType.getElementType(), baseIntType, loc, os);
+      })
+      .Case<StructType>([&](StructType structType) {
+        os << "struct packed {";
+        llvm::interleave(
+            structType.getElements(), os,
+            [&](auto &element) {
+              baseType(element.type, "logic", loc, os);
+              os << ' ';
+              printArraySubscriptsPre(element.type, loc, os);
+              os << ' ' << element.name;
+              printArraySubscriptsPost(element.type, os);
```

Comment:
Packed structs do cannot contain unpacked arrays or anything for that matter. If you encounter this situation, an error is more appropriate than printing invalid SV.

Reply:
This should be handled in the verifier, which is the point of this issue: https://github.com/llvm/circt/issues/522

Reply:
If it is handled in the verifier, there's no reason for the post subscript call. You can/should detect that case here and assert that it doesn't happen. Also, bear in mind that the RTL type system is (and will continue to be) an open type system, so moving inout and uarray into the SV dialect won't solve this issue.

Reply:
While that's true, this is ExportVerilog which has a closed world of things it understands.